blocks/basic/include/gnuradio-4.0/basic/SyncBlock.hpp

```diff
@@ -0,0 +1,363 @@
+        }
+
+        // Collect all unique sync times, ordered
+        std::set<std::uint64_t> allSyncTimes;
+        for (const auto& syncTimes : syncTimesPerInSpan) {
+            allSyncTimes.insert(syncTimes.begin(), syncTimes.end());
+        }
+        // Find the earliest sync time present in all input spans within tolerance
+        auto findInAllIt = std::ranges::find_if(allSyncTimes, [&](const auto& curTime) { //
+            return std::ranges::all_of(syncTimesPerInSpan, [&](const auto& syncTimes) {  //
+                return std::ranges::any_of(syncTimes, [&](const auto& t) { return isTimeDifferenceWithinTolerance(curTime, t); });
+            });
+        });
+
+        return findInAllIt != allSyncTimes.end() ? *findInAllIt : std::numeric_limits<std::uint64_t>::max();
+    }
+
+    [[nodiscard]] constexpr std::size_t getAvailablePreSamples(const InputSpanLike auto& in, std::size_t syncIndex) {
+        // Check if there’s an earlier sync tag that cannot be synchronized; if so, calculate available samples only up to that tag.
+        const auto foundTag = std::ranges::find_if(in.rawTags, [&](const auto& tag) { return getRelativeTagIndex(in, tag) < syncIndex && isSyncTag(tag); });
+        if (foundTag != in.rawTags.end()) {
+            return syncIndex - getRelativeTagIndex(in, *foundTag) - 1;
+        }
+        return syncIndex;
+    }
+
+    [[nodiscard]] constexpr std::size_t getAvailablePostSamples(const InputSpanLike auto& in, std::size_t syncIndex) {
+        // Check if there’s a later sync tag; if so, calculate available samples only up to that tag.
+        auto foundTag = std::ranges::find_if(in.rawTags, [&](const auto& tag) {
+            const std::size_t relativeTagIndex = getRelativeTagIndex(in, tag);
+            return relativeTagIndex > syncIndex && relativeTagIndex < in.size() && isSyncTag(tag);
+        });
+        if (foundTag != in.rawTags.end()) {
+            return getRelativeTagIndex(in, *foundTag) - syncIndex - 1;
+        }
+        return in.size() - syncIndex - 1;
+    }
+
+    [[nodiscard]] std::size_t getNSamplesBeforeSyncTag(const InputSpanLike auto& in) const {
+        for (const auto& tag : in.rawTags) {
+            const std::size_t relativeTagIndex = getRelativeTagIndex(in, tag);
+            if (isSyncTag(tag) && relativeTagIndex != std::numeric_limits<std::size_t>::max()) {
+                return std::min(relativeTagIndex, in.size()); // tags are ordered, return distance to the first sync tag
+            }
+        }
+        return in.size();
+    }
+
+    [[nodiscard]] bool isTimeDifferenceWithinTolerance(std::uint64_t t1, std::uint64_t t2) { return ((t1 > t2) ? t1 - t2 : t2 - t1) < tolerance; }
+
+    [[nodiscard]] constexpr bool isSyncTag(const gr::Tag& tag) const {
+        const std::string keyTriggerName = gr::tag::TRIGGER_NAME.shortKey();
+        const std::string keyTriggerTime = gr::tag::TRIGGER_TIME.shortKey();
```

Comment:
Why are these needed as a `std::string`? Doesn't the `at(..)` below work with string_views`

Reply:
`std::map::at` will not work with `std::sting_view`